include/zephyr/net/net_l2_connectivity.h

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for managing L2 layers that have separated post-admin-up connect
+ *	  or associate phases
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_L2_CONNECTIVITY_H_
+#define ZEPHYR_INCLUDE_NET_L2_CONNECTIVITY_H_
+
+#include <zephyr/device.h>
+#include <zephyr/net/net_if.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Network Layer 2 connectivity management abstraction layer
+ * @defgroup net_l2_connectivity Network L2 Connectivity Abstraction Layer
+ * @ingroup networking
+ * @{
+ */
+
+/**
+ * @brief Network L2 connectivity API structure
+ *
+ * Used to provide an interface to connect parameters and procedures
+ */
+struct net_if_conn;
+struct net_l2_conn_api {
+	int (*connect)(const struct net_if_conn *if_conn);
+	int (*disconnect)(const struct net_if_conn *if_conn);
+	int (*set_opt)(const struct net_if_conn *if_conn, int optname,
+							  const void *optval,
+							  size_t optlen);
+	int (*get_opt)(const struct net_if_conn *if_conn, int optname,
+							  const void *optval,
+							  size_t *optlen);
+};
+
+/**
+ * @brief L2 Connectivity Implementation struct
+ *	  Binds a connectivity implementation API to a defined L2 connectivity implementation.
+ */
+struct net_l2_conn {
+	/* The L2 connectivity implementation associated with the network device */
+	struct net_l2_conn_api *api;
+};
+
+/**
+ * @brief Network interface connectivity binding structure
+ *
+ * Binds L2 connectivity implementation to an iface / network device.
+ * Stores per-iface state for the connectivity implementation.
+ */
+struct net_if_conn {
+	/* The network interface the connectivity implementation is bound to */
+	struct net_if *iface;
+
+	/* The connectivity implementation the network device is bound to */
+	const struct net_l2_conn *impl;
+
+	/* Pointer to private, per-iface connectivity context */
+	void *ctx;
+
+	/* Generic connectivity state */
+	bool persistence;
```

Comment:
Would be helpful with some additional docs here. Especially that the timeout is in seconds.